code/datums/http.dm

```diff
@@ -0,0 +1,152 @@
+/**
+  * # HTTP Request
+  *
+  * Holder datum for ingame HTTP requests
+  *
+  * Holds information regarding to methods used, URL, and response,
+  * as well as job IDs and progress tracking for async requests
+  */
+/datum/http_request
+	/// The ID of the request (Only set if it is an async request)
+	var/id
+	/// Is the request in progress? (Only set if it is an async request)
+	var/in_progress = FALSE
+	/// HTTP method used
+	var/method
+	/// Body of the request being sent
+	var/body
+	/// Request headers being sent
+	var/headers
+	/// URL that the request is being sent to
+	var/url
+	/// The raw response, which will be decoeded into a [/datum/http_response]
+	var/_raw_response
+	/// Callback for executing after async requests. Will be called with an argument of [/datum/http_response] as first argument
+	var/datum/callback/cb
+
+/*
+###########################################################################
+THE METHODS IN THIS FILE ARE TO BE USED BY THE SUBSYSTEM AS A MANGEMENT HUB
+----------------------- DO NOT MANUALLY INVOKE THEM -----------------------
+###########################################################################
+*/
+
+/**
+  * Preparation handler
+  *
+  * Call this with relevant parameters to form the request you want to make
+  *
+  * Arguments:
+  * * _method - HTTP Method to use, see code/__DEFINES/rust_g.dm for a full list
+  * * _url - The URL to send the request to
+  * * _body - The body of the request, if applicable
+  * * _headers - Associative list of HTTP headers to send, if applicab;e
+  */
+/datum/http_request/proc/prepare(_method, _url, _body = "", list/_headers)
+	if(!length(_headers))
+		_headers = ""
+	else
+		_headers = json_encode(_headers)
+
+	method = _method
+	url = _url
+	body = _body
+	headers = _headers
```

Comment:
```suggestion
		headers = ""
	else
		headers = json_encode(_headers)

	method = _method
	url = _url
	body = _body
```
Id just assign the `headers` up top. Else you make a list variable have a string as value.

Reply:
Ill get this sorted when I get time to